### ISSUE TITLE
The confusion about results of 3DSSD between official and MMDet3D implementation.

### ISSUE THREAD

**Issue by Physu** [open-mmlab/mmdetection3d#612]
Thanks for developers extraordinary work!
I have a question about 3DSSD evaluation result between author and MMDet3D implementation.
The author's release result: [](https://github.com/dvlab-research/3DSSD)

Methods | Easy AP | Moderate AP | Hard AP | Models
-- | -- | -- | -- | --
3DSSD | 91.71 | 83.30 | 80.44 | model
PointRCNN | 88.91 | 79.88 | 78.37 | model

In MMDet3D, the result:

Backbone | Class | Lr schd | Mem (GB) | Inf time (fps) | mAP | Download
-- | -- | -- | -- | -- | -- | --
PointNet2SAMSG | Car | 72e | 4.7 | &nbsp; | 78.39(81.00)1 | model&nbsp;\|&nbsp;log

I notice "Experiment details on KITTI datasets", which shows the difference between official implementation.

1.Official implementation based on Tensorflow1.4, but I guess pytorch is not the reason of poor performance, or tensorflow and pytorch exist performance gap?
2.It is about two percent margin(81.0 and 83.3) between two implementation, can we come up with some methods to fix it?

I also use single2080Ti to train a train+val model with configs/3DSSD/3dssd_kitti-3d-car.py, I modified the `ann_file=data_root + 'kitti_infos_train.pkl',` to `ann_file=data_root + 'kitti_infos_trainval.pkl',`  the rest code was kept as origin.
when the train was finished, I will evaluate on test, and get the result there to discuss.
Thanks again!

**Comment by Physu** [comment]
I train the 3DSSD followed the configs in configs/3dssd/3dssd_kitti-3d-car.py
and modified the batchsize from 4 to 8, modified the lr from 0.002 to 0.004, the rest keep as origin.
The result(under AP40):


Benchmark | Easy | Moderate | Hard
-- | -- | -- | --
Car (Detection) | 94.91 % | 91.35 % | 87.47 %
Car (Orientation) | 0.01 % | 0.47 % | 0.63 %
Car (3D Detection) | 86.06 % | 76.48 % | 69.71 %
Car (Bird's Eye View) | 91.65 % | 86.69 % | 81.05 %



There exits a large margin between the official 3DSSD(76.48 vs 79.55). I feel confused about this, did I set something wrong? Or what can I do to make up this performance gap?
Thanks

**Comment by Physu** [comment]
Thanks for your feedback! Official code was implemented by Tensorflow, I will try train a model and submit the result to the test and evaluate the performance. New results will be updated here, as soon as I get it.

**Comment by Physu** [comment]
By the way, 79.26 is evaluated on val data or test data? If result was evaluated on test data, 79.26 vs 79.55(official in test data), the margin is acceptable. My result on test data was 3 mAP margin, it is unacceptable.

> Actually, we only reproduce the 79.26 mAP with the official code according to the record of @encore-zhou

**Comment by Tai-Wang** [comment]
It's evaluated on their val dataset and with their evaluation code (compared with the reported 83.3). So I guess there is a large range of fluctuation in terms of performance on the validation set. You can have a try first and let's have a closer look into whether there is a gap between our implementation and the official one.

**Comment by Physu** [comment]
Got it, I will try to reproduce result by following official code.

**Comment by Physu** [comment]
I use the official implementation and configs train models in Docker container.
The python packages are listed below:
tensorflow             1.4.0
tensorflow-tensorboard 0.4.0
python 3.5 
cuda 9.0
numpy 1.14.5


total train iterations: 80700
final ckptfile: model-79893(not 80700 as final ckptfile)

the result:
the model-79893
![image](https://user-images.githubusercontent.com/32857082/122037704-6d5f7f00-ce07-11eb-897c-54aea2786965.png)
the model-79086
![image](https://user-images.githubusercontent.com/32857082/122037866-9849d300-ce07-11eb-9b5b-c3a02e7874a6.png)
the model-78279
![image](https://user-images.githubusercontent.com/32857082/122037961-b1528400-ce07-11eb-9b55-8b686e13e5f8.png)
the model-77472
![image](https://user-images.githubusercontent.com/32857082/122038014-be6f7300-ce07-11eb-9f6f-9b46808743ed.png)




Benchmark | iterations | Easy | Moderate | Hard
-- | -- | -- | -- | --
Car (Detection) | 77472| 89.70 % | 82.84 % | 79.97 % 
Car (Detection) | 78279| 89.29 % | 82.69 % | 80.06 % 
Car (Detection) | 79086| 91.14 % | 82.79% | 80.02 % 
Car (Detection) | 79893| 89.39 % | 82.54% | 79.83 % 




the result is

**Comment by Tai-Wang** [comment]
It's a little strange because when we reproduce 3DSSD, @encore-zhou only got the following performance with the official code:
![image](https://user-images.githubusercontent.com/30491025/122519029-8e60e380-d044-11eb-9f3d-750c01b83c4d.png)

Maybe there is some fluctuation in performance?

**Comment by Physu** [comment]
Maybe the author improved the code implementation? There is something cause the performance gap. I will check the 3DSSD head, hope can find something to explain this situation.
![image](https://user-images.githubusercontent.com/32857082/122532971-a0964e00-d053-11eb-8ce5-b3744f2aa4d6.png)

And this is new results gotten by minutes ago.

**Comment by Physu** [comment]
By the way, this results is trained with more epoch, can see that the performance further improved( reach 82.9%).

**Comment by Tai-Wang** [comment]
Yes, it is really strange because we reproduce the above results on Aug. 2020 (as shown in the screenshot) and there are no updates after April 2020. We will check this issue recently. In the meantime, if you have any progress, please feel free to share it here.

**Comment by Physu** [comment]
Thanks for reopen this issue! New finding will be updated.

**Comment by Physu** [comment]
Using pytorch1.5
mmdet 1.3.9
mmdet3d 0.14.0
mmcv-full 1.3.9
ubuntu 18.4

I use official configs in `configs/3DSSD/3dssd_4x4_kitti-3d-car.py` and modified the batchsize from 4 to 8, lr_rate keep as origin.
Trained a model with 2 2080Ti GPUs and with Full train data(7481). Finally I get a validation results on valsplit (3, 769 samples):
![image](https://user-images.githubusercontent.com/32857082/125248739-04e0c080-e327-11eb-883e-ba9aa74bffbd.png)
Then I generate test submission file, and submit it to test server:
![image](https://user-images.githubusercontent.com/32857082/125248913-3c4f6d00-e327-11eb-880b-d172bb17d5c8.png)
The performance is not good as I expected, I just don't know why?

**Comment by Physu** [comment]
I find it is hard to reproduce the results on KITTI test, though you could have gotten a good result on val already.

**Comment by Physu** [comment]
If we set the confidence threshold great than 0.0, e.g. 0.2 to filter the final predictions in test.txt, we will get ![image](https://user-images.githubusercontent.com/32857082/125716135-1cc1999d-e7ce-413b-85f5-50caf207e263.png). 
Note that in configs, you can define your threshold:
```
test_cfg=dict(
        nms_cfg=dict(type='nms', iou_thr=0.1),
        sample_mod='spec',
        score_thr=0.0,  # Attention!!!
        per_class_proposal=True,
        max_output_num=100))
```
Though there is some improvement, it is far from 79.57(3DSSD in leaderboard). I guess a good post processing is needed，but the other skills which can improve performance are sitll a question.

**Comment by Wuziyi616** [comment]
@Physu Have you ever tried generating submission using the **official** code and submit it to the test server to see the test set result? Also, it seems to me that, changing mmdet3d's training batch and GPUs from `4x4` to `8x2` improves val set results a lot?

**Comment by Physu** [comment]
@Wuziyi616 Thanks for your attention! Does **offcial code** mean [dvlab-research/3DSSD](https://github.com/dvlab-research/3DSSD) or other methods?

**Comment by Physu** [comment]
I will reproduce on 4*4, and we will see the difference further.

**Comment by Wuziyi616** [comment]
> 
> 
> @Wuziyi616 Thanks for your attention! Does **offcial code** mean [dvlab-research/3DSSD](https://github.com/dvlab-research/3DSSD) or other methods?
> Besides, in order to learn more about the evaluation procedure, I use [traveller59/kitti-object-eval-python](https://github.com/traveller59/kitti-object-eval-python) to test results on val set(e.g. get every LiDAR bin's results and save it in a txt file, finally get 3769 txt files).
> I find when no other post processing involved, the results:
> ![image](https://user-images.githubusercontent.com/32857082/125887141-8c5ef216-8cb1-4ccf-92c7-37f687f1b76a.png)
> which is slightly better than mmdet3d evaluation result(Maybe it is unfair to compare this way, for the hyperparameters may change):
> ![image](https://user-images.githubusercontent.com/32857082/125887196-f0787a3f-0df0-4c56-8174-6e7a56d9ab72.png)
> If I use a confidence threshold 0.2 to filter out the false positive, the result further improved:
> ![image](https://user-images.githubusercontent.com/32857082/125887470-2dd681dd-c2f4-483b-bdf0-3ac4fe23f214.png)

Exactly, the official code I said is the dvlab's code. I think that's the official code release for 3DSSD isn't it? As you mentioned in [this](https://github.com/open-mmlab/mmdetection3d/issues/612#issuecomment-856598907) reply, you said you would like to submit test results using that code, have you done that?

**Comment by Wuziyi616** [comment]
@Physu Hi would you mind adding me on Wechat so we can use Chines to better discuss this problem? I find some strange lines in our implementation and want to discuss it with you. My email is dazitu616@gmail.com, you can send your Wechat account to me if you want :)

**Comment by Physu** [comment]
Thanks for your attention，my opportunity is running out, the results will be updated soon.

**Comment by jlqzzz** [comment]
@Physu 
Have you tried to reproduce the multi-class version of 3dssd (that is, predict car, pedestrian and cyclist at the same time)?

**Comment by Machine97** [comment]
@Physu 
Hi, have you ever tried generating submission using the official code and submit it to the test server to see the test set result?
